manifests/agent/config.pp

```diff
@@ -1,91 +1,62 @@
 class puppet::agent::config {
 
-  include puppet
-  include puppet::agent
-  include puppet::params
-
   Ini_setting {
-    path    => $puppet::params::puppet_conf,
+    path    => $puppet::conf,
+    section => 'agent',
     ensure  => 'present',
   }
 
-  ini_setting { 'server':
-    section => 'main',
-    setting => 'server',
-    value   => $puppet::agent::server,
+  if $puppet::agent::ca_server {
+    $real_ca_server = $puppet::agent::ca_server
+  } else {
+    $real_ca_server = $puppet::agent::server
   }
 
-  ini_setting { 'ca_server':
-    section => 'main',
-    setting => 'ca_server',
-    value   => $puppet::agent::real_ca_server,
+  if $puppet::agent::report_server {
+    $real_report_server = $puppet::agent::report_server
+  } else {
+    $real_report_server = $puppet::agent::server
   }
 
-  ini_setting { 'report_server':
-    section => 'main',
-    setting => 'report_server',
-    value   => $puppet::agent::real_report_server,
-  }
-
-  if $puppet::agent::report_format {
-    ini_setting { 'report_format':
+  ini_setting {
+    'server':
       section => 'main',
-      setting => 'report_format',
-      value   => $puppet::agent::report_format,
-    }
-  }
-
-  ini_setting { 'pluginsync':
-    section => 'agent',
-    setting => 'pluginsync',
-    value   => $puppet::agent::pluginsync,
-  }
-
-  ini_setting { 'certname':
-    section => 'agent',
-    setting => 'certname',
-    value   => $puppet::agent::certname,
-  }
-
-  ini_setting { 'report':
-    section => 'agent',
-    setting => 'report',
-    value   => $puppet::agent::report,
-  }
-
-  ini_setting { 'environment':
-    section => 'agent',
-    setting => 'environment',
-    value   => $puppet::agent::environment,
-  }
-
-  ini_setting { 'show_diff':
-    section => 'agent',
-    setting => 'show_diff',
-    value   => $puppet::agent::showdiff,
-  }
-
-  ini_setting { 'splay':
-    section => 'agent',
-    setting => 'splay',
-    value   => $puppet::agent::splay,
-  }
-
-  ini_setting { 'configtimeout':
-    section => 'agent',
-    setting => 'configtimeout',
-    value   => $puppet::agent::configtimeout,
-  }
-
-  ini_setting { 'usecacheonfailure':
-    section => 'agent',
-    setting => 'usecacheonfailure',
-    value   => $puppet::agent::usecacheonfailure,
-  }
-
-  ini_setting { 'stringify_facts_agent':
-    setting => 'stringify_facts',
-    section => 'agent',
-    value   => $puppet::agent::stringify_facts;
+      setting => 'server',
+      value   => $puppet::agent::server;
+    'ca_server':
```

Comment:
If the 'main' options truly belong in `[main]` than perhaps we can simplify by putting them in the `pupept` class instead of the `agent` class.  Then overrides get simpler.  Just an idea.